add runbook for ExtremelyHighIndividualControlPlaneCPU (#29)
openshift/runbooks: alerts/cluster-kube-apiserver-operator/ExtremelyHighIndividualControlPlaneCPU.md

# ExtremelyHighIndividualControlPlaneCPU

## Meaning

> [These alerts][ExtremelyHighIndividualControlPlaneCPU] are triggered when

Reviewer: ```suggestion
[This alert][ExtremelyHighIndividualControlPlaneCPU] is triggered when
```

```

These are the conditions that could trigger the alert:

- there is a new workload that is generating more calls to the apiserver
  and causing high CPU usage. In this case, increase the CPU and
  memory on your control plane nodes.
- the alert is triggered based on the node metrics, so it could be that a
  component on the node is causing the high CPU usage.
- apiserver/etcd is processing more requests due to client retries that is
  being caused by an underlying condition.
- uneven distribution of requests to the apiserver instance(s) due to http2
  (it multiplexes requests over a single TCP connection). The load balancers
  are not at application layer, and so does not understand http2.

## Mitigation

- if a workload is generating load to the apiserver that is causing high CPU
  usage, then increase the CPU and memory on your control plane nodes.
- If the sustained high CPU usage is due to a cluster degradation:

  - find out the root cause of the degradation, and
  - then determine the next steps accordingly.

If this needs to be reported, then capture the following dataset, and file
a new issue in BugZilla with links to the captured dataset

- must-gather
- audit logs

> - dump of prometheus data

Author: prom dump is needed (especially in this case) when reporter wants to create a BZ. Unfortunately, we have not yet productized  prom dump yet. this script is all we have for now to capture prom data.

# ExtremelyHighIndividualControlPlaneCPU

## Meaning

> [This alert][ExtremelyHighIndividualControlPlaneCPU](https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/bindata/assets/alerts/cpu-utilization.yaml)

Reviewer: Preview is broken here, you should change that to:
```suggestion
[This alert][ExtremelyHighIndividualControlPlaneCPU]
```
and add the following at the end of the file:
```
[ExtremelyHighIndividualControlPlaneCPU]: https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/bindata/assets/alerts/cpu-utilization.yaml
```

Same applies for `HighOverallControlPlaneCPU`.

is triggered when CPU utilization across all three control plane nodes is
higher than two control plane nodes can sustain; a single control plane
node outage may cause a cascading failure; increase available CPU.

The urgency of this alert is determined by how long CPU utilization across all
three control plane nodes is higher than two control plane nodes can sustain.

* Warning
  * when CPU utilization across all three control plane nodes is higher than
    two control plane nodes can sustain for more than `10m`.
  

## Impact

Extreme CPU pressure can cause slow serialization and poor performance from
the `kube-apiserver` and `etcd`. When this happens, there is a risk of
clients seeing non-responsive API requests which are issued again
causing even more CPU pressure.

It can also cause failing liveness probes due to slow etcd responsiveness on
the backend. If one kube-apiserver fails under this condition, chances are
you will experience a cascade as the remaining kube-apiservers
are also under-provisioned.

To fix this, increase the CPU and memory on your control plane nodes.

## Diagnosis

The following prometheus queries can be used to diagnose:
```sh

> // CPU utilization of container(s) running on a particular node

Reviewer: ```suggestion
// Top 5 of containers with the most CPU utilization on a particular node
```

The urgency of this alert is determined by how long CPU utilization across all
three control plane nodes is higher than two control plane nodes can sustain.

* Warning
  * when CPU utilization across all three control plane nodes is higher than
    two control plane nodes can sustain for more than `10m`.
  

## Impact

Extreme CPU pressure can cause slow serialization and poor performance from
the `kube-apiserver` and `etcd`. When this happens, there is a risk of
clients seeing non-responsive API requests which are issued again
causing even more CPU pressure.

It can also cause failing liveness probes due to slow etcd responsiveness on
the backend. If one kube-apiserver fails under this condition, chances are
you will experience a cascade as the remaining kube-apiservers
are also under-provisioned.

To fix this, increase the CPU and memory on your control plane nodes.

## Diagnosis

The following prometheus queries can be used to diagnose:
```sh
// CPU utilization of container(s) running on a particular node
topk(5,
  sum by (namespace, pod, container) (
    irate (container_cpu_usage_seconds_total{node="NODE_NAME",container!="",pod!=""}[5m])
  )
)

> // CPU utilization on master nodes

Reviewer: ```suggestion
// CPU utilization of containers on master nodes
```

sum by (node) (
  irate (container_cpu_usage_seconds_total{container!="",pod!=""}[5m])
    and on (node) cluster:master_nodes
)

> // inspect cgroup

Reviewer: ```suggestion
// CPU utilization of the master nodes from the cgroups
```

# ExtremelyHighIndividualControlPlaneCPU

## Meaning

> [This alert][ExtremelyHighIndividualControlPlaneCPU]is triggered when there

Reviewer: ```suggestion
[This alert][ExtremelyHighIndividualControlPlaneCPU] is triggered when there
```
